src/bmf.cpp

```diff
@@ -467,193 +459,153 @@ long int CxadbmfPlayer::__bmf_convert_stream(const unsigned char *stream,
+      event.cmd = 0xFE;
       if (stream_end - stream < 2)
         return -1;
-      bmf.streams[channel][pos].cmd = 0xFE;
-      bmf.streams[channel][pos].cmd_data = (*(stream+1) & ((bmf.version == BMF0_9B) ? 0x7F : 0x3F)) - 1;
-
-	  stream+=2;
-    }
-    else if (*stream == 0x7D)
-	{
-      // 0x7D -> 0xFD: Loop Saved Position
-      bmf.streams[channel][pos].cmd = 0xFD;
-
-	  stream++;
-	}
-	else
-    {
-      if (*stream & 0x80)                           
-      {
-        if (stream_end - stream < 2)
-          return -1;
-		if (*(stream+1) & 0x80)
-        {
-		  if (*(stream+1) & 0x40)
-          {
-            // byte0: 1aaaaaaa = NOTE
-            bmf.streams[channel][pos].note = *stream & 0x7F;
-            // byte1: 11bbbbbb = DELAY
-            bmf.streams[channel][pos].delay = *(stream+1) & 0x3F;
-            // byte2: cccccccc = COMMAND
-
-            stream+=2;
+      event.cmd_data = (stream[1] & (bmf.version == BMF0_9B ? 0x7F : 0x3F)) - 1;
+      stream += 2;
+      break;
 
-            is_cmd = true;
-          }
-		  else
-          {
-            // byte0: 1aaaaaaa = NOTE
-            bmf.streams[channel][pos].note = *stream & 0x7F;
-            // byte1: 11bbbbbb = DELAY
-            bmf.streams[channel][pos].delay = *(stream+1) & 0x3F;
-
-			stream+=2;
-          } // if (*(stream+1) & 0x40)
-		}
-        else
-        {
-          // byte0: 1aaaaaaa = NOTE
-          bmf.streams[channel][pos].note = *stream & 0x7F;
-          // byte1: 0bbbbbbb = COMMAND
+    case 0x7D: // 0x7D -> 0xFD: Loop to Saved Position
+      event.cmd = 0xFD;
+      stream++;
+      break;
 
-          stream++;
+      /* Normal Events */
+    default:
+      // The event is defined by the next 1 to 4 bytes and consists of
+      // a note, an optional delay, and an optional command (which may
+      // take another data byte as argument).
+      // Bit 0x80 of the first byte indicates the presence of an optional
+      // part; if present, bits 0x80 and 0x40 of the second byte determine
+      // whether we have delay, command, or both. The possible forms are:
+      //   0NNNNNNN                   - 1  byte:  note
+      //   1NNNNNNN 0CCCCCCC          - 2+ bytes: note, command [data]
+      //   1NNNNNNN 10DDDDDD          - 2  bytes: note, delay
+      //   1NNNNNNN 11DDDDDD CCCCCCCC - 3+ bytes: note, delay, command [data]
+      event.note = *stream & 0x7F;
+
+      if (!(*(stream++) & 0x80))
+        break; // 1st byte is 0NNNNNNN: no delay or command; done.
 
-          is_cmd = true;
-        } // if (*(stream+1) & 0x80)
-	  }
-	  else
+      if (stream_end - stream < 1)
+        return -1;
+      if (*stream & 0x80)
       {
-        // byte0: 0aaaaaaa = NOTE
-        bmf.streams[channel][pos].note = *stream & 0x7F;
+        // 2nd byte is 1xDDDDDD: delay byte present
+        event.delay = *stream & 0x3F;
 
-		stream++;
-      } // if (*stream & 0x80)
-    } // if (*stream == 0xFE)
+        if (!(*(stream++) & 0x40))
```

Comment:
Please confirm but I believe you can get away without the brackets here and in the other places: `*stream++` (as the `++` doesn't apply until the whole thing has finished being evaluated).

Reply:
Yes, the parentheses can be omitted since postfix operators have higher precedence than prefix unary operators, but for some people that's hard to memorize. I like the added clarity of the patentheses, but don't like the cluttered display, so it's a mixed bag, again.

Reply:
You make a fair point, but by the same token memorising the operator precedence is probably going to serve someone well in the long run :)